cmake/external/mkldnn.cmake

```diff
@@ -0,0 +1,78 @@
+# Copyright (c) 2017 PaddlePaddle Authors. All Rights Reserve.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+IF(NOT ${WITH_MKLDNN})
+  return()
+ENDIF(NOT ${WITH_MKLDNN})
+
+INCLUDE(ExternalProject)
+
+SET(MKLDNN_PROJECT "extern_mkldnn")
+SET(MKLDNN_SOURCES_DIR ${THIRD_PARTY_PATH}/mkldnn)
+SET(MKLDNN_INSTALL_DIR ${THIRD_PARTY_PATH}/install/mkldnn)
+SET(MKLDNN_INCLUDE_DIR "${MKLDNN_INSTALL_DIR}/include" CACHE PATH "mkldnn include directory." FORCE)
+
+# The following magic numbers should be updated regularly to keep latest version
+SET(MKLDNN_TAG "v0.9")
+SET(MKLDNN_MKL_VER "mklml_lnx_2018.0.20170425")
+
+IF(WIN32)
+    MESSAGE(WARNING "It is not supported compiling with mkldnn in windows Paddle yet."
+      "Force WITH_MKLDNN=OFF")
+    SET(WITH_MKLDNN OFF)
+    return()
+ELSE(WIN32)
+    SET(MKLDNN_LIBRARY "${MKLDNN_INSTALL_DIR}/lib/libmkldnn.so" CACHE FILEPATH "mkldnn library." FORCE)
+    MESSAGE(STATUS "Set ${MKLDNN_INSTALL_DIR}/lib to runtime path")
+    SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+    #SET(CMAKE_MACOSX_RPATH 1) # hold for MacOS
+    SET(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_RPATH}" "${MKLDNN_INSTALL_DIR}/lib")
+ENDIF(WIN32)
+
+INCLUDE_DIRECTORIES(${MKLDNN_INCLUDE_DIR})
+
+SET(MKLDNN_CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+SET(MKLDNN_CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+
+ExternalProject_Add(
+    ${MKLDNN_PROJECT}
+    ${EXTERNAL_PROJECT_LOG_ARGS}
+    GIT_REPOSITORY    "https://github.com/01org/mkl-dnn.git"
+    GIT_TAG           "${MKLDNN_TAG}"
+    PREFIX            ${MKLDNN_SOURCES_DIR}
+    PATCH_COMMAND     cd <SOURCE_DIR>/scripts && ./prepare_mkl.sh
```

Comment:
The similar error when I `sh prepare_mkl.sh`:
```
[luotao02@yq01-idl-gpu-jpaas-let01: scripts] (tags/v0.9) -> $ outnet_exec sh prepare_mkl.sh 
--2017-07-19 18:37:52--  https://github.com/01org/mkl-dnn/releases/download/v0.9/mklml_lnx_2018.0.20170425.tgz
Resolving agent.baidu.com... 172.22.15.16, 172.22.15.17
Connecting to agent.baidu.com|172.22.15.16|:8118... connected.
Proxy request sent, awaiting response... 302 Found
Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/58414589/54f3715e-3c9a-11e7-97f0-b402ecc00a80?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20170719%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170719T103753Z&X-Amz-Expires=300&X-Amz-Signature=7a6e4b2a2bf0f14ff8916d77c9244e67c339220e21e9970d0f0cb20af016f3d5&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dmklml_lnx_2018.0.20170425.tgz&response-content-type=application%2Foctet-stream [following]
--2017-07-19 18:37:53--  https://github-production-release-asset-2e65be.s3.amazonaws.com/58414589/54f3715e-3c9a-11e7-97f0-b402ecc00a80?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20170719%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170719T103753Z&X-Amz-Expires=300&X-Amz-Signature=7a6e4b2a2bf0f14ff8916d77c9244e67c339220e21e9970d0f0cb20af016f3d5&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dmklml_lnx_2018.0.20170425.tgz&response-content-type=application%2Foctet-stream
Connecting to agent.baidu.com|172.22.15.16|:8118... connected.
Proxy tunneling failed: ForbiddenUnable to establish SSL connection.
tar (child): /home/luotao02/Paddle/third_party/mkldnn/src/extern_mkldnn/external/mklml_lnx*.tgz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
Downloaded and unpacked Intel(R) MKL small libraries to /home/luotao02/Paddle/third_party/mkldnn/src/extern_mkldnn/external
``` 
As the scripts not return `False` when download fails, thus `make` continuous, and will encounter:
```
In file included from /home/luotao02/Paddle/paddle/function/GemmFunctor.h:17:0,
                 from /home/luotao02/Paddle/paddle/function/GemmConvOp.cpp:16:
/home/luotao02/Paddle/paddle/math/MathFunctions.h:19:23: fatal error: mkl_cblas.h: No such file or directory
 #include <mkl_cblas.h>
                       ^
compilation terminated.
```